dan_layer/wallet/sdk/src/apis/jwt.rs

```diff
@@ -0,0 +1,205 @@
+                "StartWebrtc" => Ok(JrpcPermission::StartWebrtc),
+                "Admin" => Ok(JrpcPermission::Admin),
+                _ => Err(InvalidJrpcPermissionsFormat(s.to_string())),
+            },
+        }
+    }
+}
+
+#[derive(Debug, Clone, Serialize, Deserialize)]
+pub struct JrpcPermissions(pub Vec<JrpcPermission>);
+
+impl FromStr for JrpcPermissions {
+    type Err = InvalidJrpcPermissionsFormat;
+
+    fn from_str(s: &str) -> Result<Self, Self::Err> {
+        Ok(JrpcPermissions(
+            s.split(',').map(JrpcPermission::from_str).collect::<Result<_, _>>()?,
+        ))
+    }
+}
+
+impl JrpcPermissions {
+    pub fn no_permissions(&self) -> bool {
+        self.0.is_empty()
+    }
+
+    pub fn has_permission(&self, permission: &JrpcPermission) -> bool {
+        self.0.contains(permission) || self.0.contains(&JrpcPermission::Admin)
+    }
+}
+
+#[derive(Debug, Serialize, Deserialize)]
+struct Claims {
+    id: u64,
+    permissions: JrpcPermissions,
+    exp: usize,
+}
+
+// This is used when you request permission.
+#[derive(Debug, Serialize, Deserialize)]
+pub struct AuthClaims {
+    id: u64,
+    permissions: JrpcPermissions,
+    exp: usize,
+}
+
+impl<'a, TStore: WalletStore> JwtApi<'a, TStore> {
+    pub(crate) fn new(store: &'a TStore, expiration: Duration, secret_key: String) -> Self {
+        Self {
+            store,
+            expiration,
+            auth_secret_key: format!("auth-{secret_key}"),
+            jwt_secret_key: format!("jwt-{secret_key}"),
+        }
+    }
+
+    // Get and also increment index. We could probably use random id here.
+    pub fn get_index(&self) -> Result<u64, JwtApiError> {
+        let mut tx = self.store.create_write_tx()?;
+        let index = tx.jwt_add_empty_token()?;
+        tx.commit()?;
+        Ok(index)
+    }
+
+    pub fn generate_auth_token(&self, permissions: JrpcPermissions) -> Result<String, JwtApiError> {
+        let id = self.get_index()?;
+
+        let my_claims = AuthClaims {
+            id,
+            permissions,
+            exp: (SystemTime::now() + self.expiration)
+                .duration_since(UNIX_EPOCH)
+                .unwrap()
+                .as_secs() as usize,
+        };
+        let auth_token = encode(
+            &Header::default(),
+            &my_claims,
+            &EncodingKey::from_secret(self.auth_secret_key.as_ref()),
+        )?;
+        Ok(auth_token)
+    }
+
+    pub fn check_auth_token(&self, auth_token: &str) -> Result<AuthClaims, JwtApiError> {
+        let auth_token_data = decode::<AuthClaims>(
+            auth_token,
+            &DecodingKey::from_secret(self.auth_secret_key.as_ref()),
+            &Validation::default(),
+        )?;
+        Ok(auth_token_data.claims)
+    }
+
+    pub fn get_permissions(&self, token: Option<String>) -> Result<JrpcPermissions, JwtApiError> {
+        let token_data = decode::<Claims>(
+            token.ok_or(JwtApiError::TokenMissing)?.as_str(),
```

Comment:
nit: Rather make it the caller's responsibility to provide a token
```suggestion
    pub fn get_permissions(&self, token: &str) -> Result<JrpcPermissions, JwtApiError> {
        let token_data = decode::<Claims>(
            token
```

Reply:
I wanted to, but then I will have to do the check somewhere else (in the jrpc_server handler function).

One solution (I don't know if it's possible) is to have two handlers (one with token and one without), then I could also get rid of the axum::middleware for the token.

Reply:
Yeah I think they should be checked elsewhere - like the `check_auth` I wrote

Reply:
I also want the correct json error to be returned (see `check_auth`)

Reply:
Ah now I understand what you did. Ok that works though I prefer less extra type handling code in the generic error resolver, but happy with the approach in any case. I think that the `check_auth` function should then live in jwt_api and should return `Result<(), JwtApiError>`